src/rgw/driver/rados/rgw_bucketlogging.cc

```diff
@@ -0,0 +1,605 @@
+#include "include/function2.hpp"
+#include "rgw_sal_rados.h"
+#include "rgw_zone_features.h"
+#include "rgw_perf_counters.h"
+#include "services/svc_zone.h"
+#include "common/dout.h"
+#include "rgw_url.h"
+#include <chrono>
+#include <fmt/format.h>
+#include "librados/AioCompletionImpl.h"
+#include "common/async/yield_waiter.h"
+#include <future>
+#include <thread>
+
+#define dout_subsys ceph_subsys_rgw_bucket_logging
+
+namespace rgw::bucket_logging {
+
+using commit_list_t = std::set<std::string>;
+
+// use mmap/mprotect to allocate 128k coroutine stacks
+auto make_stack_allocator() {
+  return boost::context::protected_fixedsize_stack{128*1024};
+}
+
+const std::string COMMIT_LIST_OBJECT_NAME = "bucket_logging_commit_list";
+
+class BucketLoggingManager : public DoutPrefixProvider {
+  using Executor = boost::asio::io_context::executor_type;
+  bool m_shutdown = false;
+  static constexpr auto m_log_commits_update_period = std::chrono::milliseconds(30000); // 30s
+  static constexpr auto m_log_commits_update_retry = std::chrono::milliseconds(1000); // 1s
+  static constexpr auto m_log_commits_idle_sleep = std::chrono::milliseconds(100); // 100ms
+  const utime_t m_failover_time = utime_t(m_log_commits_update_period*3); // 90s
+  CephContext* const m_cct;
+  static constexpr auto COOKIE_LEN = 16;
+  const std::string m_lock_cookie;
+  boost::asio::io_context m_io_context;
+  boost::asio::executor_work_guard<Executor> m_work_guard;
+  std::thread m_worker;
+  const SiteConfig& m_site;
+  rgw::sal::RadosStore* const m_rados_store;
+
+private:
+
+  CephContext *get_cct() const override { return m_cct; }
+
+  unsigned get_subsys() const override { return dout_subsys; }
+
+  std::ostream& gen_prefix(std::ostream& out) const override {
+    return out << "rgw bucket_logging: ";
+  }
+
+  int parse_list_name(std::string list_obj_name, std::string &tenant_name,
+                      std::string &bucket_name, std::string &bucket_id,
+                      std::string &prefix) {
+    // <tenant_name>/<bucket_name>/<bucket_id>/<prefix>
+    size_t pstart = 0, pend = 0;
+    pend =list_obj_name.find('/');
+    if (pend == std::string::npos) {
+      return -EINVAL;
+    }
+    tenant_name = list_obj_name.substr(pstart, pend);
+
+    pstart = pend + 1;
+    pend =list_obj_name.find('/', pstart);
+    if (pend == std::string::npos) {
+      return -EINVAL;
+    }
+    bucket_name = list_obj_name.substr(pstart, pend - pstart);
+    if (bucket_name.empty()) {
+    return -EINVAL;
+    }
+
+    pstart = pend + 1;
+    pend =list_obj_name.find('/', pstart);
+    if (pend == std::string::npos) {
+      return -EINVAL;
+    }
+    bucket_id = list_obj_name.substr(pstart, pend - pstart);
+    if (bucket_id.empty()) {
+      return -EINVAL;
+    }
+    prefix = list_obj_name.substr(pend + 1);
+    return 0;
+  }
+
+  // read the list of commit lists from the global list object
+  int read_commit_lists(commit_list_t& commits, optional_yield y) {
```

Comment:
we probably want to rename this function to reflect that it reads a global list and not list of commits